azure/scope/machine.go

```diff
@@ -289,17 +292,35 @@ func (m *MachineScope) DiskSpecs() []azure.ResourceSpecGetter {
 }
 
 // RoleAssignmentSpecs returns the role assignment specs.
-func (m *MachineScope) RoleAssignmentSpecs() []azure.RoleAssignmentSpec {
+func (m *MachineScope) RoleAssignmentSpecs(principalID *string) []azure.ResourceSpecGetter {
+	roles := make([]azure.ResourceSpecGetter, 1)
 	if m.AzureMachine.Spec.Identity == infrav1.VMIdentitySystemAssigned {
-		return []azure.RoleAssignmentSpec{
-			{
-				MachineName:  m.Name(),
-				Name:         m.AzureMachine.Spec.RoleAssignmentName,
-				ResourceType: azure.VirtualMachine,
-			},
+		roles[0] = &roleassignments.RoleAssignmentSpec{
+			Name:             m.AzureMachine.Spec.RoleAssignmentName,
+			MachineName:      m.Name(),
+			ResourceType:     azure.VirtualMachine,
+			ResourceGroup:    m.ResourceGroup(),
+			Scope:            azure.GenerateScope(m.SubscriptionID()),
+			RoleDefinitionID: azure.GenerateContributorRoleDefinitionID(m.SubscriptionID()),
+			PrincipalID:      principalID,
 		}
+		fmt.Println(roles[0])
+		return roles
 	}
-	return []azure.RoleAssignmentSpec{}
+	return []azure.ResourceSpecGetter{}
+}
+
+// RoleAssignmentResourceType returns the role assignment resource type.
+func (m *MachineScope) RoleAssignmentResourceType() string {
+	if m.AzureMachine.Spec.Identity == infrav1.VMIdentitySystemAssigned {
+		return azure.VirtualMachine
+	}
+	return ""
+}
+
+// MachineName returns the machine name.
+func (m *MachineScope) MachineName() string {
```

Comment:
why not just use `m.Name()`?

Reply:
This is actually an implementation of `RoleAssignmentScope` interface method. 

See this https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1873/files#diff-5480fcd7b8b0c3331bc1e5049598731653f0133de8a39be5553d1b6e49a0026dR43

I had to do this because -- the `Service` has the following: 
```
// Service provides operations on Azure resources.
type Service struct {
	Scope RoleAssignmentScope
	client
	async.Reconciler
	virtualMachinesClient        virtualmachines.Client
	virtualMachineScaleSetClient scalesets.Client
}
```
Now this `Scope` in the `Service` struct is an interface of type `RoleAssignmentScope`. So to get access of machine name, I had to implement the method. Other way would have been a cast of the value but that seemed not appropriate. Open to suggestions.

Reply:
And machine name is required to fetch the VM for principal ID in the reconcile block.